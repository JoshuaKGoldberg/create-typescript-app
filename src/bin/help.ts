import chalk from "chalk";

import { allArgOptions } from "../shared/options/args.js";

interface HelpTextSection {
	sectionHeading: string;
	subsections: {
		flags: SubsectionFlag[];
		subheading?: string;
		warning?: string;
	}[];
}

interface SubsectionFlag {
	description: string;
	flag: string;
	type: string;
}

function logHelpTextSection(section: HelpTextSection): void {
	console.log(" ");

	console.log(`${chalk.black.bgGreenBright(section.sectionHeading)}`);

	for (const subsection of section.subsections) {
		if (subsection.warning) {
			console.log(chalk.yellow(subsection.warning));
		}

		if (subsection.subheading) {
			console.log(chalk.green(subsection.subheading));
		}

		for (const { description, flag, type } of subsection.flags) {
			console.log(
				chalk.cyan(
					`
  --${flag}${
		type !== "boolean" ? ` (${chalk.cyanBright(type)})` : ""
	}: ${description}`,
				),
			);
		}
	}
}

function createHelpTextSections(): HelpTextSection[] {
	const core: HelpTextSection = {
		sectionHeading: "Core options:",
		subsections: [
			{
				flags: [],
			},
		],
	};

	const optional: HelpTextSection = {
		sectionHeading: "Optional options:",
		subsections: [
			{
				flags: [],
			},
		],
	};

	const optOut: HelpTextSection = {
		sectionHeading: "Opt-outs:",
		subsections: [
			{
				flags: [],
				warning: `
  ⚠️ Warning: Specifying any --exclude-* flag on the command-line will 
  cause the setup script to skip prompting for more excludes. ⚠️`,
			},
			{
				flags: [
					{
						description: `Skips network calls that fetch all-contributors
  data from GitHub`,
						flag: "exclude-contributors",
						type: "boolean",
					},
				],
				subheading: `
You can prevent the migration script from making some network-based 
changes using any or all of the following CLI flags:`,
			},
			{
				flags: [],
				subheading: `
You can prevent the migration script from making some changes on disk 
using any or all of the following CLI flags:`,
			},
		],
	};

	const offline: HelpTextSection = {
		sectionHeading: "Offline Mode:",
		subsections: [
			{
				flags: [],
			},
		],
	};

	const subsections = {
		core: core.subsections[0],
		offline: offline.subsections[0],
		"opt-out": optOut.subsections[0],
		optional: optional.subsections[0],
		"skip-disk": optOut.subsections[2],
		"skip-net": optOut.subsections[1],
	};

	for (const [option, data] of Object.entries(allArgOptions)) {
		subsections[data.docsSection].flags.push({
			description: data.description,
			flag: option,
			type: data.type,
		});
	}

	return [core, optional, optOut, offline];
}

export function logHelpText(introLogs: string[]): void {
	const helpTextSections = createHelpTextSections();

	for (const log of introLogs) {
		console.log(log);
		console.log(" ");
	}

	console.log(
		chalk.cyan(
			`
A quickstart-friendly TypeScript template with comprehensive formatting, 
linting, releases, testing, and other great tooling built-in.
      `,
		),
	);

	for (const section of helpTextSections) {
		logHelpTextSection(section);

		console.log();
	}
}
