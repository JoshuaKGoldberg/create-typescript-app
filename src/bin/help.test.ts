import chalk from "chalk";

import { describe, expect, it, SpyInstance, vi, beforeEach } from "vitest";
import { logHelpText } from "./help.js";

let mockConsoleLog: SpyInstance;

describe("logHelpText", () => {
	beforeEach(() => {
		mockConsoleLog = vi
			.spyOn(console, "log")
			.mockImplementation(() => undefined);
	});

	it("logs help text when called", async () => {
		logHelpText([
			chalk.yellow(
				"⚠️ This template is early stage, opinionated, and not endorsed by the TypeScript team. ⚠️",
			),
			chalk.yellow(
				"⚠️ If any tooling it sets displeases you, you can always remove that portion manually. ⚠️",
			),
		]);

		expect(mockConsoleLog.mock.calls).toMatchSnapshot();
	});
});
