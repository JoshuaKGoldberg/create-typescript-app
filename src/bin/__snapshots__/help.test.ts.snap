// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logHelpText > logs help text when called 1`] = `
[
  [
    "[33m‚ö†Ô∏è This template is early stage, opinionated, and not endorsed by the TypeScript team. ‚ö†Ô∏è[39m",
  ],
  [
    " ",
  ],
  [
    "[33m‚ö†Ô∏è If any tooling it sets displeases you, you can always remove that portion manually. ‚ö†Ô∏è[39m",
  ],
  [
    " ",
  ],
  [
    "[36m[39m
[36mA quickstart-friendly TypeScript template with comprehensive formatting, [39m
[36mlinting, releases, testing, and other great tooling built-in.[39m
[36m      [39m",
  ],
  [
    " ",
  ],
  [
    "[30m[102mCore options:[49m[39m",
  ],
  [
    "[36m[39m
[36m  --base ([96mstring[39m[36m): Whether to scaffold the repository with:[39m
[36m  ‚Ä¢ everything: that comes with the template ([96m[1mrecommended[22m[39m[36m)[39m
[36m  ‚Ä¢ minimum: amounts of tooling, essentially opting out of everything[39m
[36m  ‚Ä¢ prompt: for which portions to exclude[39m",
  ],
  [
    "[36m[39m
[36m  --create-repository: Whether to create a corresponding repository on [39m
[36m  github.com (if it doesn't yet exist)[39m",
  ],
  [
    "[36m[39m
[36m  --description ([96mstring[39m[36m): Sentence case description of the repository [39m
[36m  (e.g. A quickstart-friendly TypeScript package with lots of great [39m
[36m  repository tooling. ‚ú®)[39m",
  ],
  [
    "[36m[39m
[36m  --mode ([96mstring[39m[36m): Whether to:[39m
[36m  ‚Ä¢ create: a new repository in a child directory[39m
[36m  ‚Ä¢ initialize: a freshly repository in the current directory[39m
[36m  ‚Ä¢ migrate: an existing repository in the current directory[39m",
  ],
  [
    "[36m[39m
[36m  --owner ([96mstring[39m[36m): GitHub organization or user the repository is [39m
[36m  underneath (e.g. JoshuaKGoldberg)[39m",
  ],
  [
    "[36m[39m
[36m  --repository ([96mstring[39m[36m): The kebab-case name of the repository [39m
[36m  (e.g. create-typescript-app)[39m",
  ],
  [
    "[36m[39m
[36m  --title ([96mstring[39m[36m): Title Case title for the repository to be used in [39m
[36m  documentation (e.g. Create TypeScript App)[39m",
  ],
  [],
  [
    " ",
  ],
  [
    "[30m[102mOptional options:[49m[39m",
  ],
  [
    "[36m[39m
[36m  --access ([96mstring[39m[36m): ([96m\\"public\\" | \\"restricted\\"[39m[36m): Which [96mnpm publish --access[39m[36m to [39m
[36m  release npm packages with (by default, \\"public\\")[39m",
  ],
  [
    "[36m[39m
[36m  --author ([96mstring[39m[36m): Username on npm to publish packages under (by [39m
[36m  default, an existing npm author, or the currently logged in npm user, [39m
[36m  or [96mowner.toLowerCase()[39m[36m)[39m",
  ],
  [
    "[36m[39m
[36m  --directory ([96mstring[39m[36m): Directory to create the repository in (by default, the same [39m
[36m  name as the repository)[39m",
  ],
  [
    "[36m[39m
[36m  --email ([96mstring[39m[36m): Email address to be listed as the point of contact in [39m
[36m  docs and packages (e.g. example@joshuakgoldberg.com)[39m",
  ],
  [
    "[36m[39m
[36m  --email-github ([96mstring[39m[36m): Optionally, may be provided to use different emails in [96m.md[39m[36m [39m
[36m  files[39m",
  ],
  [
    "[36m[39m
[36m  --email-npm ([96mstring[39m[36m): Optionally, may be provided to use different emails in [39m
[36m  [96mpackage.json[39m[36m[39m",
  ],
  [
    "[36m[39m
[36m  --funding ([96mstring[39m[36m): GitHub organization or username to mention in [39m
[36m  [96mfunding.yml[39m[36m (by default, owner)[39m",
  ],
  [
    "[36m[39m
[36m  --keywords ([96mstring[39m[36m): Any number of keywords to include in [96mpackage.json[39m[36m (by default, [39m
[36m  none). This can be specified any number of times, like  [39m
[36m  [96m--keywords apple --keywords \\"banana cherry\\"[39m[36m[39m",
  ],
  [
    "[36m[39m
[36m  --logo ([96mstring[39m[36m): Local image file in the repository to display near the top of [39m
[36m  the README.md as a logo[39m",
  ],
  [
    "[36m[39m
[36m  --logo-alt ([96mstring[39m[36m): If [96m--logo[39m[36m is provided or detected from an existing README.md, [39m
[36m  alt text that describes the image will be prompted for if not provided[39m",
  ],
  [
    "[36m[39m
[36m  --preserve-generated-form: Whether to keep the GitHub repository generated from notice[39m
[36m  (by default, false)[39m",
  ],
  [],
  [
    " ",
  ],
  [
    "[30m[102mOpt-outs:[49m[39m",
  ],
  [
    "[33m[39m
[33m  ‚ö†Ô∏è Warning: Specifying any --exclude-* flag on the command-line will [39m
[33m  cause the setup script to skip prompting for more excludes. ‚ö†Ô∏è[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-all-contributors: Don't add all-contributors to track contributions [39m
[36m  and display them in a README.md table.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-compliance: Don't add a GitHub Actions workflow to verify that [39m
[36m  PRs match an expected format.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-deprecation: Don't use eslint-plugin-deprecation to report on usage of [39m
[36m  code marked as [96m@deprecated[39m[36m.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-jsdoc: Don't use eslint-plugin-jsdoc to enforce good practices around [39m
[36m  JSDoc comments.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-json: Don't apply linting and sorting to [96m*.json[39m[36m, and [39m
[36m  [96m*.jsonc[39m[36m files.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-knip: Don't add Knip to detect unused files, dependencies, [39m
[36m  and code exports.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-md: Don't apply linting to [96m*.md[39m[36m files.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-package-json: Don't add npm-package-json-lint to lint for [39m
[36m  package.json correctness.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-packages: Don't add a pnpm dedupe workflow to ensure [39m
[36m  packages aren't duplicated unnecessarily.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-perfectionist: Don't apply eslint-plugin-perfectionist to [39m
[36m  ensure imports, keys, and so on are in sorted order.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-regex: Don't add eslint-plugin-regex to enforce good practices around [39m
[36m  regular expressions.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-spelling: Don't add cspell to spell check against [39m
[36m  dictionaries of known words.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-strict: Don't augment the recommended logical lint rules with [39m
[36m  typescript-eslint's strict config.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-stylistic: Don't add stylistic rules such as typescript-eslint's [39m
[36m  stylistic config.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-lint-yml: Don't apply linting and sorting to [96m*.yaml[39m[36m [39m
[36m  and [96m*.yml[39m[36m files.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-releases: Don't add release-it to generate changelogs, package [39m
[36m  bumps, and publishes based on conventional commits.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-renovate: Don't add a Renovate config to dependencies up-to-date with [39m
[36m  PRs.[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-tests: Don't add Vitest tooling for fast unit tests, configured [39m
[36m  with coverage tracking.[39m",
  ],
  [
    "[32m[39m
[32mYou can prevent the migration script from making some network-based [39m
[32mchanges using any or all of the following CLI flags:[39m",
  ],
  [
    "[36m[39m
[36m  --exclude-contributors: Skips network calls that fetch all-contributors[39m
[36m  data from GitHub[39m",
  ],
  [
    "[36m[39m
[36m  --skip-all-contributors-api: Skips network calls that fetch all-contributors data from[39m
[36m  GitHub. This flag does nothing if [96m--exclude-all-contributors[39m[36m was specified.[39m",
  ],
  [
    "[36m[39m
[36m  --skip-github-api: Skips calling to GitHub APIs.[39m",
  ],
  [
    "[36m[39m
[36m  --skip-install: Skips installing all the new template [39m
[36m  packages with pnpm.[39m",
  ],
  [
    "[32m[39m
[32mYou can prevent the migration script from making some changes on disk [39m
[32musing any or all of the following CLI flags:[39m",
  ],
  [
    "[36m[39m
[36m  --skip-removal: Skips removing setup docs and scripts, [39m
[36m  including this [96mdocs/[39m[36m directory[39m",
  ],
  [
    "[36m[39m
[36m  --skip-restore: Skips the prompt offering to restore the [39m
[36m  repository if an error occurs during setup[39m",
  ],
  [
    "[36m[39m
[36m  --skip-uninstall: Skips uninstalling packages only used for [39m
[36m  setup scripts[39m",
  ],
  [],
  [
    " ",
  ],
  [
    "[30m[102mOffline Mode:[49m[39m",
  ],
  [
    "[36m[39m
[36m  --offline: You can run [96mcreate-typescript-app[39m[36m in an \\"offline\\" mode. [39m
[36m  Doing so will:[39m
[36m  ‚Ä¢ Enable [96m--exclude-all-contributors-api[39m[36m and [96m--skip-github-api[39m[36m[39m
[36m  ‚Ä¢ Skip network calls when setting up contributors[39m
[36m  ‚Ä¢ Run pnpm commands with pnpm's [96m--offline[39m[36m mode[39m",
  ],
  [],
]
`;
